# $$Electric-vehicle-data-analysis$$
---

In [36]:
import numpy as np
import pandas as pd
import matplotlib as plt
import seaborn as sb 
import os

In [37]:
working_dir = os.getcwd()
data_1 = pd.read_csv(open(f'{working_dir}\\datasets\\electric_vehicle_sales_by_state.csv', 'rb'))
data_1

date                     state vehicle_category  \
0     01-Apr-21                    Sikkim       2-Wheelers   
1     01-Apr-21                    Sikkim       4-Wheelers   
2     01-May-21                    Sikkim       2-Wheelers   
3     01-May-21                    Sikkim       4-Wheelers   
4     01-Jun-21                    Sikkim       2-Wheelers   
...         ...                       ...              ...   
2440  01-Mar-24                   Mizoram       2-Wheelers   
2441  01-Mar-24                DNH and DD       2-Wheelers   
2442  01-Mar-24                   Manipur       2-Wheelers   
2443  01-Mar-24  Andaman & Nicobar Island       2-Wheelers   
2444  01-Mar-24                  Nagaland       2-Wheelers   

      electric_vehicles_sold  total_vehicles_sold  
0                          0                  398  
1                          0                  361  
2                          0                  113  
3                          0                   98  
4                          0                  229  
...                      ...                  ...  
2440                      58                 1932  
2441                      25                  780  
2442                      13                 1394  
2443                       2                  447  
2444                       2                 1180  

[2445 rows x 5 columns]

In [38]:
data_1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2445 entries, 0 to 2444
Data columns (total 5 columns):
 #   Column                  Non-Null Count  Dtype 
---  ------                  --------------  ----- 
 0   date                    2445 non-null   object
 1   state                   2445 non-null   object
 2   vehicle_category        2445 non-null   object
 3   electric_vehicles_sold  2445 non-null   int64 
 4   total_vehicles_sold     2445 non-null   int64 
dtypes: int64(2), object(3)
memory usage: 95.6+ KB


<b>

- Dataframe size: 2445 * 5

- No null values present in the dataframe

- Only date column datatype is not appropriate; means we need to change it into datetime format

In [39]:
data_1['date'] = pd.to_datetime(data_1['date'],format='%d-%b-%y')

In [40]:
data_1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2445 entries, 0 to 2444
Data columns (total 5 columns):
 #   Column                  Non-Null Count  Dtype         
---  ------                  --------------  -----         
 0   date                    2445 non-null   datetime64[ns]
 1   state                   2445 non-null   object        
 2   vehicle_category        2445 non-null   object        
 3   electric_vehicles_sold  2445 non-null   int64         
 4   total_vehicles_sold     2445 non-null   int64         
dtypes: datetime64[ns](1), int64(2), object(2)
memory usage: 95.6+ KB


-----
---

In [41]:
data_1.nunique()

date                        36
state                       35
vehicle_category             2
electric_vehicles_sold     921
total_vehicles_sold       2225
dtype: int64

In [42]:
working_dir = os.getcwd()
data_2 = pd.read_csv(open(f'{working_dir}\\datasets\\electric_vehicle_sales_by_makers.csv', 'rb'))
data_2

date vehicle_category            maker  electric_vehicles_sold
0    01-Apr-21       2-Wheelers     OLA ELECTRIC                       0
1    01-Apr-22       2-Wheelers         OKAYA EV                       0
2    01-May-21       2-Wheelers     OLA ELECTRIC                       0
3    01-Jun-21       2-Wheelers     OLA ELECTRIC                       0
4    01-Jul-21       2-Wheelers     OLA ELECTRIC                       0
..         ...              ...              ...                     ...
811  01-Mar-24       2-Wheelers           BGAUSS                    3070
812  01-Mar-24       2-Wheelers  BATTRE ELECTRIC                     625
813  01-Mar-24       2-Wheelers    KINETIC GREEN                    3915
814  01-Mar-24       2-Wheelers           REVOLT                     585
815  01-Mar-24       2-Wheelers           OTHERS                   10579

[816 rows x 4 columns]

In [43]:
data_2.nunique()

date                       36
vehicle_category            2
maker                      26
electric_vehicles_sold    603
dtype: int64

In [44]:
data_2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 816 entries, 0 to 815
Data columns (total 4 columns):
 #   Column                  Non-Null Count  Dtype 
---  ------                  --------------  ----- 
 0   date                    816 non-null    object
 1   vehicle_category        816 non-null    object
 2   maker                   816 non-null    object
 3   electric_vehicles_sold  816 non-null    int64 
dtypes: int64(1), object(3)
memory usage: 25.6+ KB


<b>

- Dataframe size: 816 * 4

- No null values present in the dataframe

- Only date column datatype is not appropriate; means we need to change it into datetime format

In [45]:
data_2['date'] = pd.to_datetime(data_2['date'],format='%d-%b-%y')

In [46]:
data_2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 816 entries, 0 to 815
Data columns (total 4 columns):
 #   Column                  Non-Null Count  Dtype         
---  ------                  --------------  -----         
 0   date                    816 non-null    datetime64[ns]
 1   vehicle_category        816 non-null    object        
 2   maker                   816 non-null    object        
 3   electric_vehicles_sold  816 non-null    int64         
dtypes: datetime64[ns](1), int64(1), object(2)
memory usage: 25.6+ KB


---
---

In [47]:
working_dir = os.getcwd()
data_3 = pd.read_csv(open(f'{working_dir}\\datasets\\dim_date.csv', 'rb'))
data_3

date  fiscal_year quarter
0   01-Apr-21         2022      Q1
1   01-May-21         2022      Q1
2   01-Jun-21         2022      Q1
3   01-Jul-21         2022      Q2
4   01-Aug-21         2022      Q2
5   01-Sep-21         2022      Q2
6   01-Oct-21         2022      Q3
7   01-Nov-21         2022      Q3
8   01-Dec-21         2022      Q3
9   01-Jan-22         2022      Q4
10  01-Feb-22         2022      Q4
11  01-Mar-22         2022      Q4
12  01-Apr-22         2023      Q1
13  01-May-22         2023      Q1
14  01-Jun-22         2023      Q1
15  01-Jul-22         2023      Q2
16  01-Aug-22         2023      Q2
17  01-Sep-22         2023      Q2
18  01-Oct-22         2023      Q3
19  01-Nov-22         2023      Q3
20  01-Dec-22         2023      Q3
21  01-Jan-23         2023      Q4
22  01-Feb-23         2023      Q4
23  01-Mar-23         2023      Q4
24  01-Apr-23         2024      Q1
25  01-May-23         2024      Q1
26  01-Jun-23         2024      Q1
27  01-Jul-23         2024      Q2
28  01-Aug-23         2024      Q2
29  01-Sep-23         2024      Q2
30  01-Oct-23         2024      Q3
31  01-Nov-23         2024      Q3
32  01-Dec-23         2024      Q3
33  01-Jan-24         2024      Q4
34  01-Feb-24         2024      Q4
35  01-Mar-24         2024      Q4

In [50]:
data_3.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 36 entries, 0 to 35
Data columns (total 3 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   date         36 non-null     object
 1   fiscal_year  36 non-null     int64 
 2   quarter      36 non-null     object
dtypes: int64(1), object(2)
memory usage: 996.0+ bytes


In [51]:
data_3['date'] = pd.to_datetime(data_3['date'],format='%d-%b-%y')

In [52]:
data_3.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 36 entries, 0 to 35
Data columns (total 3 columns):
 #   Column       Non-Null Count  Dtype         
---  ------       --------------  -----         
 0   date         36 non-null     datetime64[ns]
 1   fiscal_year  36 non-null     int64         
 2   quarter      36 non-null     object        
dtypes: datetime64[ns](1), int64(1), object(1)
memory usage: 996.0+ bytes


In [55]:
state_data = pd.merge(data_1, data_3,on='date',how='left')
maker_data = pd.merge(data_2, data_3,on='date',how='left')

In [58]:
state_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2445 entries, 0 to 2444
Data columns (total 7 columns):
 #   Column                  Non-Null Count  Dtype         
---  ------                  --------------  -----         
 0   date                    2445 non-null   datetime64[ns]
 1   state                   2445 non-null   object        
 2   vehicle_category        2445 non-null   object        
 3   electric_vehicles_sold  2445 non-null   int64         
 4   total_vehicles_sold     2445 non-null   int64         
 5   fiscal_year             2445 non-null   int64         
 6   quarter                 2445 non-null   object        
dtypes: datetime64[ns](1), int64(3), object(3)
memory usage: 133.8+ KB


In [59]:
maker_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 816 entries, 0 to 815
Data columns (total 6 columns):
 #   Column                  Non-Null Count  Dtype         
---  ------                  --------------  -----         
 0   date                    816 non-null    datetime64[ns]
 1   vehicle_category        816 non-null    object        
 2   maker                   816 non-null    object        
 3   electric_vehicles_sold  816 non-null    int64         
 4   fiscal_year             816 non-null    int64         
 5   quarter                 816 non-null    object        
dtypes: datetime64[ns](1), int64(2), object(3)
memory usage: 38.4+ KB


---

## Data Inspection for state_data

In [60]:
state_data.head()

date   state vehicle_category  electric_vehicles_sold  \
0 2021-04-01  Sikkim       2-Wheelers                       0   
1 2021-04-01  Sikkim       4-Wheelers                       0   
2 2021-05-01  Sikkim       2-Wheelers                       0   
3 2021-05-01  Sikkim       4-Wheelers                       0   
4 2021-06-01  Sikkim       2-Wheelers                       0   

   total_vehicles_sold  fiscal_year quarter  
0                  398         2022      Q1  
1                  361         2022      Q1  
2                  113         2022      Q1  
3                   98         2022      Q1  
4                  229         2022      Q1

In [66]:
state_data.isna().sum()

date                      0
state                     0
vehicle_category          0
electric_vehicles_sold    0
total_vehicles_sold       0
fiscal_year               0
quarter                   0
dtype: int64

In [61]:
state_data.nunique()

date                        36
state                       35
vehicle_category             2
electric_vehicles_sold     921
total_vehicles_sold       2225
fiscal_year                  3
quarter                      4
dtype: int64

In [62]:
state_data['date'].unique()

<DatetimeArray>
['2021-04-01 00:00:00', '2021-05-01 00:00:00', '2021-06-01 00:00:00',
 '2021-07-01 00:00:00', '2021-08-01 00:00:00', '2021-09-01 00:00:00',
 '2021-10-01 00:00:00', '2021-11-01 00:00:00', '2021-12-01 00:00:00',
 '2022-01-01 00:00:00', '2022-02-01 00:00:00', '2022-03-01 00:00:00',
 '2022-04-01 00:00:00', '2022-05-01 00:00:00', '2022-06-01 00:00:00',
 '2022-07-01 00:00:00', '2022-08-01 00:00:00', '2022-09-01 00:00:00',
 '2022-10-01 00:00:00', '2022-11-01 00:00:00', '2022-12-01 00:00:00',
 '2023-01-01 00:00:00', '2023-02-01 00:00:00', '2023-03-01 00:00:00',
 '2023-04-01 00:00:00', '2023-05-01 00:00:00', '2023-06-01 00:00:00',
 '2023-07-01 00:00:00', '2023-08-01 00:00:00', '2023-09-01 00:00:00',
 '2023-10-01 00:00:00', '2023-11-01 00:00:00', '2023-12-01 00:00:00',
 '2024-01-01 00:00:00', '2024-02-01 00:00:00', '2024-03-01 00:00:00']
Length: 36, dtype: datetime64[ns]

In [63]:
state_data['state'].unique()

array(['Sikkim', 'Andaman & Nicobar Island', 'Arunachal Pradesh', 'Assam',
       'Chhattisgarh', 'DNH and DD', 'Jammu and Kashmir', 'Ladakh',
       'Manipur', 'Meghalaya', 'Mizoram', 'Nagaland', 'Puducherry',
       'Tripura', 'Bihar', 'Chandigarh', 'Delhi', 'Madhya Pradesh',
       'Odisha', 'Punjab', 'Uttarakhand', 'Himachal Pradesh',
       'Andaman & Nicobar', 'Haryana', 'Jharkhand', 'Andhra Pradesh',
       'Goa', 'Gujarat', 'Karnataka', 'Kerala', 'Maharashtra',
       'Rajasthan', 'Tamil Nadu', 'Uttar Pradesh', 'West Bengal'],
      dtype=object)

In [64]:
state_data['vehicle_category'].unique()

array(['2-Wheelers', '4-Wheelers'], dtype=object)

---

#### Identify the top 5 states with the highest penetration rate in 2-wheeler and 4-wheeler EV sales in FY 2024

In [90]:
state_data[state_data['date'].dt.year == 2024]

date                     state vehicle_category  \
66   2024-01-01                    Sikkim       2-Wheelers   
67   2024-01-01                    Sikkim       4-Wheelers   
68   2024-02-01                    Sikkim       2-Wheelers   
69   2024-02-01                    Sikkim       4-Wheelers   
70   2024-03-01                    Sikkim       2-Wheelers   
...         ...                       ...              ...   
2440 2024-03-01                   Mizoram       2-Wheelers   
2441 2024-03-01                DNH and DD       2-Wheelers   
2442 2024-03-01                   Manipur       2-Wheelers   
2443 2024-03-01  Andaman & Nicobar Island       2-Wheelers   
2444 2024-03-01                  Nagaland       2-Wheelers   

      electric_vehicles_sold  total_vehicles_sold  fiscal_year quarter  
66                         0                  509         2024      Q4  
67                         0                  422         2024      Q4  
68                         0                  408         2024      Q4  
69                         0                  375         2024      Q4  
70                         0                  399         2024      Q4  
...                      ...                  ...          ...     ...  
2440                      58                 1932         2024      Q4  
2441                      25                  780         2024      Q4  
2442                      13                 1394         2024      Q4  
2443                       2                  447         2024      Q4  
2444                       2                 1180         2024      Q4  

[202 rows x 7 columns]